src/middlewared/middlewared/plugins/auth_/authenticate.py

```diff
@@ -0,0 +1,71 @@
+import crypt
+import hmac
+
+from middlewared.service import Service, private
+
+
+class AuthService(Service):
+
+    class Config:
+        cli_namespace = 'auth'
+
+    @private
+    async def authenticate(self, username, password):
+        if '@' in username:
+            if (await self.middleware.call('datastore.config', 'system.settings'))['stg_ds_auth']:
+                return await self.ds_authenticate(username, password)
+            else:
+                return None
+        else:
+            return await self.local_authenticate(username, password)
+
+    @private
+    async def local_authenticate(self, username, password):
+        try:
+            user = await self.middleware.call(
+                'datastore.query',
+                'account.bsdusers',
+                [('bsdusr_username', '=', username)],
+                {'get': True},
+            )
+        except IndexError:
+            return None
+
+        if user['bsdusr_unixhash'] in ('x', '*'):
+            return None
+
+        if not hmac.compare_digest(crypt.crypt(password, user['bsdusr_unixhash']), user['bsdusr_unixhash']):
+            return None
+
+        return await self.authenticate_local_user(user['id'], username)
+
+    @private
+    async def authenticate_local_user(self, user_id, username):
+        gids = {
+            member['bsdgrpmember_group']['bsdgrp_gid']
+            for member in await self.middleware.call(
+                'datastore.query',
+                'account.bsdgroupmembership',
+                [['bsdgrpmember_user', '=', user_id]],
+            )
+        }
+
+        return await self.common_authenticate(username, 'local_groups', gids)
+
+    @private
+    async def ds_authenticate(self, username, password):
+        return None  # FIXME
```

Comment:
If you are just needing a list of groups for a uid, you can do:
```
root@truenas[~]# midclt call user.get_user_obj '{"uid": 1000, "get_groups": true}' | jq
{
  "pw_name": "smbuser",
  "pw_uid": 1000,
  "pw_gid": 1000,
  "pw_gecos": "smbuser",
  "pw_dir": "/nonexistent",
  "pw_shell": "/usr/bin/bash",
  "grouplist": [
    1000,
    545
  ]
}
```
This works for all users (local and directory services).